File: tools/testing/python/testing_server.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS-IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Tink Primitive Testing Service in Python."""

from __future__ import absolute_import
from __future__ import division
# Placeholder for import for type annotations
from __future__ import print_function

from concurrent import futures
import io

from absl import app
from absl import flags
import grpc
import tink
from tink import aead
from tink import cleartext_keyset_handle
from tools.testing import tink_testing_pb2
from tools.testing import tink_testing_pb2_grpc
from google3.net.grpc.python import loas2
from tink.proto import tink_pb2

FLAGS = flags.FLAGS

flags.DEFINE_integer('port', 10000, 'The port of the server.')


class TinkTestingServicer(tink_testing_pb2_grpc.TinkTestingServicer):
  """A Tink Primitive Testing gRPC server."""

  def GenerateKeyset(
      self, request: tink_testing_pb2.GenerateKeysetRequest,
      context: grpc.ServicerContext) -> tink_testing_pb2.GenerateKeysetResponse:
    """Generates a keyset."""
    try:
      template = tink_pb2.KeyTemplate()
      template.ParseFromString(request.template)
      keyset_handle = tink.new_keyset_handle(template)
      keyset = io.BytesIO()
      cleartext_keyset_handle.write(
          tink.BinaryKeysetWriter(keyset), keyset_handle)
      return tink_testing_pb2.GenerateKeysetResponse(keyset=keyset.getvalue())
    except tink.TinkError as e:
      return tink_testing_pb2.GenerateKeysetResponse(err=str(e))

  def AeadEncrypt(
      self, request: tink_testing_pb2.AeadEncryptRequest,
      context: grpc.ServicerContext) -> tink_testing_pb2.AeadEncryptResponse:
    """Encrypts a message."""
    try:
      keyset_handle = cleartext_keyset_handle.read(
          tink.BinaryKeysetReader(request.keyset))
      p = keyset_handle.primitive(aead.Aead)
      ciphertext = p.encrypt(request.plaintext, request.associated_data)
      return tink_testing_pb2.AeadEncryptResponse(ciphertext=ciphertext)
    except tink.TinkError as e:
      return tink_testing_pb2.AeadEncryptResponse(err=str(e))

  def AeadDecrypt(
      self, request: tink_testing_pb2.AeadDecryptRequest,
      context: grpc.ServicerContext) -> tink_testing_pb2.AeadDecryptResponse:
    """Decrypts a message."""
    try:
      keyset_handle = cleartext_keyset_handle.read(
          tink.BinaryKeysetReader(request.keyset))
      p = keyset_handle.primitive(aead.Aead)
      plaintext = p.decrypt(request.ciphertext, request.associated_data)
      return tink_testing_pb2.AeadDecryptResponse(plaintext=plaintext)
    except tink.TinkError as e:
      return tink_testing_pb2.AeadDecryptResponse(err=str(e))


def main(unused_argv):
  aead.register()
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=2))
  servicer = TinkTestingServicer()
  tink_testing_pb2_grpc.add_TinkTestingServicer_to_server(servicer, server)
  server.add_secure_port('[::]:%d' % FLAGS.port,
                         loas2.loas2_server_credentials())
  server.start()
  server.wait_for_termination()


if __name__ == '__main__':
  app.run(main)

File: tools/testing/go/testing_server.go
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

// Package main is implements an gRPC server for testing_api.
package main

import (
	"bytes"
	"context"
	"fmt"
	"net"

	"flag"
	// context is used to cancel outstanding requests
	"google3/base/go/log"
	"google3/net/grpc/go/grpcprod"
	"google3/third_party/golang/grpc/grpc"
	"github.com/google/tink/go/aead"
	"github.com/google/tink/go/keyset"
	"github.com/google/tink/go/testkeyset"
	tinkpb "github.com/google/tink/go/proto/tink_go_proto"
	pbgrpc "google3/third_party/tink/tools/testing/testing_api_go_grpc"
	pb "google3/third_party/tink/tools/testing/testing_api_go_proto"

	"github.com/golang/protobuf/proto"
)

var (
	port = flag.Int("port", 10000, "The server port")
)

// KeysetService implements the Keyset testing service.
type KeysetService struct {
}

func (s *KeysetService) Generate(ctx context.Context, req *pb.GenerateKeysetRequest) (*pb.KeysetResponse, error) {
	template := &tinkpb.KeyTemplate{}
	err := proto.Unmarshal(req.Template, template)
	if err != nil {
		return &pb.KeysetResponse{
			Result: &pb.KeysetResponse_Err{err.Error()}}, nil
	}
	handle, err := keyset.NewHandle(template)
	if err != nil {
		return &pb.KeysetResponse{
			Result: &pb.KeysetResponse_Err{err.Error()}}, nil
	}
	buf := new(bytes.Buffer)
	writer := keyset.NewBinaryWriter(buf)
	err = testkeyset.Write(handle, writer)
	if err != nil {
		return &pb.KeysetResponse{
			Result: &pb.KeysetResponse_Err{err.Error()}}, nil
	}
	return &pb.KeysetResponse{
		Result: &pb.KeysetResponse_Keyset{buf.Bytes()}}, nil
}

// AeadService implements the Aead testing service.
type AeadService struct {
}

func (s *AeadService) Encrypt(ctx context.Context, req *pb.AeadEncryptRequest) (*pb.CiphertextResponse, error) {
	reader := keyset.NewBinaryReader(bytes.NewReader(req.Keyset))
	handle, err := testkeyset.Read(reader)
	if err != nil {
		return &pb.CiphertextResponse{
			Result: &pb.CiphertextResponse_Err{err.Error()}}, nil
	}
	cipher, err := aead.New(handle)
	if err != nil {
		return &pb.CiphertextResponse{
			Result: &pb.CiphertextResponse_Err{err.Error()}}, nil
	}
	ciphertext, err := cipher.Encrypt(req.Plaintext, req.AssociatedData)
	if err != nil {
		return &pb.CiphertextResponse{
			Result: &pb.CiphertextResponse_Err{err.Error()}}, nil
	}
	return &pb.CiphertextResponse{
		Result: &pb.CiphertextResponse_Ciphertext{ciphertext}}, nil
}

func (s *AeadService) Decrypt(ctx context.Context, req *pb.AeadDecryptRequest) (*pb.PlaintextResponse, error) {
	reader := keyset.NewBinaryReader(bytes.NewReader(req.Keyset))
	handle, err := testkeyset.Read(reader)
	if err != nil {
		return &pb.PlaintextResponse{
			Result: &pb.PlaintextResponse_Err{err.Error()}}, nil
	}
	cipher, err := aead.New(handle)
	if err != nil {
		return &pb.PlaintextResponse{
			Result: &pb.PlaintextResponse_Err{err.Error()}}, nil
	}
	plaintext, err := cipher.Decrypt(req.Ciphertext, req.AssociatedData)
	if err != nil {
		return &pb.PlaintextResponse{
			Result: &pb.PlaintextResponse_Err{err.Error()}}, nil
	}
	return &pb.PlaintextResponse{
		Result: &pb.PlaintextResponse_Plaintext{plaintext}}, nil
}

func main() {
	google.Init()
	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", *port))
	if err != nil {
		log.Exitf("Server failed to listen: %v", err)
	}
	log.Infof("Server is now listening on port: %d", *port)
	var serverOpts []grpc.ServerOption
	serverConfig := grpcprod.DefaultServerConfig()
	server, err := grpcprod.NewServer(serverConfig, serverOpts...)
	if err != nil {
		log.Exitf("Failed to create new grpcprod server: %v", err)
	}
	pbgrpc.RegisterKeysetServer(server.GRPC, &KeysetService{})
	pbgrpc.RegisterAeadServer(server.GRPC, &AeadService{})
	server.Serve(lis)
}
